lib/puppet-ghostbuster/puppetdb.rb

```diff
@@ -13,14 +13,22 @@ class PuppetDB
     end
 
     def self.client
-      @@client ||= ::PuppetDB::Client.new({
-                                            server: ENV['PUPPETDB_URL'] || @@puppetdb,
-                                            pem: {
-                                              'key' => ENV['PUPPETDB_KEY_FILE'] || Puppet[:hostprivkey],
-                                              'cert' => ENV['PUPPETDB_CERT_FILE'] || Puppet[:hostcert],
-                                              'ca_file' => ENV['PUPPETDB_CACERT_FILE'] || Puppet[:localcacert],
-                                            },
-                                          }, 4)
+      options = {
+        server: ENV['PUPPETDB_URL'] || @@puppetdb,
+      }
+
+      if ENV['PE_TOKEN']
+        options[:token] = ENV['PE_TOKEN']
+        options[:cacert] = ENV['PUPPETDB_CACERT_FILE'] || Puppet[:localcacert]
+      else
+        options[:pem] = {
+          'key' => ENV['PUPPETDB_KEY_FILE'] || Puppet[:hostprivkey],
+          'cert' => ENV['PUPPETDB_CERT_FILE'] || Puppet[:hostcert],
+          'ca_file' => ENV['PUPPETDB_CACERT_FILE'] || Puppet[:localcacert],
+        }
+      end
+
+      @@client ||= ::PuppetDB::Client.new(options, 4)
```

Comment:
I think you should use a block to cache. The diff may be hard to read, but it's essentially
```ruby
@@client ||= begin
  # ...
end
```

```suggestion
      @@client ||= begin
        options = {
          server: ENV['PUPPETDB_URL'] || @@puppetdb,
        }

        if ENV['PE_TOKEN']
          options[:token] = ENV['PE_TOKEN']
          options[:cacert] = ENV['PUPPETDB_CACERT_FILE'] || Puppet[:localcacert]
        else
          options[:pem] = {
            'key' => ENV['PUPPETDB_KEY_FILE'] || Puppet[:hostprivkey],
            'cert' => ENV['PUPPETDB_CERT_FILE'] || Puppet[:hostcert],
            'ca_file' => ENV['PUPPETDB_CACERT_FILE'] || Puppet[:localcacert],
          }
        end

        ::PuppetDB::Client.new(options, 4)
      end
```

Reply:
Cheers. Fixed.